src/compiler/checker.ts

```diff
@@ -9514,100 +9674,140 @@ module ts {
-                //     [Symbol.iterator]: { // iteratorFunction
-                //         (): { // iterator
-                //             next: { // iteratorNextFunction
-                //                 (): { // iteratorNextResult
-                //                     value: T // iteratorNextValue
-                //                 }
-                //             }
-                //         }
-                //     }
-                // }
-                //
-                // T is the type we are after. At every level that involves analyzing return types
-                // of signatures, we union the return types of all the signatures.
-                //
-                // Another thing to note is that at any step of this process, we could run into a dead end,
-                // meaning either the property is missing, or we run into the anyType. If either of these things
-                // happens, we return undefined to signal that we could not find the iterated type. If a property
-                // is missing, and the previous step did not result in 'any', then we also give an error if the
-                // caller requested it. Then the caller can decide what to do in the case where there is no iterated
-                // type. This is different from returning anyType, because that would signify that we have matched the
-                // whole pattern and that T (above) is 'any'.
-
-                if (allConstituentTypesHaveKind(iterable, TypeFlags.Any)) {
-                    return undefined;
-                }
+        function getElementTypeFromIterable(iterable: Type, errorNode: Node): Type {
+            // We want to treat type as an iterable, and get the type it is an iterable of. The iterable
+            // must have the following structure (annotated with the names of the variables below):
+            //
+            // { // iterable
+            //     [Symbol.iterator]: { // iteratorFunction
+            //         (): Iterator<T>
+            //     }
+            // }
+            //
+            // T is the type we are after. At every level that involves analyzing return types
+            // of signatures, we union the return types of all the signatures.
+            //
+            // Another thing to note is that at any step of this process, we could run into a dead end,
+            // meaning either the property is missing, or we run into the anyType. If either of these things
+            // happens, we return undefined to signal that we could not find the iterated type. If a property
+            // is missing, and the previous step did not result in 'any', then we also give an error if the
+            // caller requested it. Then the caller can decide what to do in the case where there is no iterated
+            // type. This is different from returning anyType, because that would signify that we have matched the
+            // whole pattern and that T (above) is 'any'.

```

Comment:
But I digressed. In this situation, we have sufficient documentation. I just meant that it is a little more clear that this is a caller-intended comment rather than an explanation of implementation.